Fix validations on child record when record parent has validate: false

eileencodes opened the pull request:
Fixes #17621. This 5 year old (or older) issue causes validations to fire when a parent record has `validate: false` option and a child record is saved. It's not the responsibility of the model to validate an associated object unless the object was created or modified by the parent.

Also cleaned up tests related to validations
`assert_nothing_raised` is not benefiting us in these tests
Corrected spelling of "respects"
It's better to use `assert_not_operator` over `assert !r.valid`

Originally @tenderlove and I had put the definition of `should_validate?` in ActiveModel but I found later that `persisted?`, `changed?` and `marked_for_destruction?` are not defined in ActiveModel so those tests were failing. I moved that code completely into AR. If there is another suggestion to a better place for this method I'm open to suggestions. 

P.S. I'm on vacation this week so I might be slow to respond if you have questions :sunny: :smile:

eileencodes committed:
Fix validations on child record when record parent has validate: false

Fixes #17621. This 5 year old (or older) issue causes validations to fire
when a parent record has `validate: false` option and a child record is
saved. It's not the responsibility of the model to validate an
associated object unless the object was created or modified by the
parent.

Clean up tests related to validations

`assert_nothing_raised` is not benefiting us in these tests
Corrected spelling of "respects"
It's better to use `assert_not_operator` over `assert !r.valid`

sgrif commented on activerecord/lib/active_record/validations/uniqueness.rb:
```diff
@@ -28,6 +29,10 @@ def validate_each(record, attribute, value)
       end
 
+      def should_validate?(record)
+        !record.persisted? || record.changed? || record.marked_for_destruction?
```
Should this be extracted somewhere, since it's re-used three times?

sgrif commented:
We might want to benchmark this, since calling `record.changed?` repeatedly will probably cause a major regression.

tenderlove commented:
It looks like `changed?` is just checking to see if a hash is empty.  Is it doing something more?  Also this change only impacts specific validations, I have a hard time believing it would cause a major speed regression in an app.

sgrif commented:
The hash calculates in-place changes.

sgrif commented:
Maybe "major" is a bit far, but I would expect it to move the needle for `Model#save` if it has multiple presence validations.

eileencodes commented:
Thanks for the feedback @sgrif - I'll consolidate the code and run benchmarks when I'm back in town this weekend.

eileencodes commented:
@sgrif - I moved the `should_validate?` definition into ActiveModel but am not calling it from there so the tests pass. That was the only shared file between the validations - if this is undesirable the only other place I can think to put it would be it's own file. :smile_cat: 

Regarding benchmarks: Maybe I have vacation brain but I'm thinking about the benchmarks for this and struggling to get something that makes sense and we can be confident in.

Technically I can't run the validation on `#save` and benchmark it before the change because it will raise `ActiveRecord::RecordInvalid` (the reason for this PR). When an error is raised the benchmark doesn't finish so I have 2 options: tack on a `validate: false` onto the second call to save or  `rescue` the `ActiveRecord::RecordInvalid`.

`validate: false` isn't going to work because we're then skipping the code section I added so it would be the same benchmark on master vs this branch. The code added in this branch is faster than a rescue but I don't think that's what we're looking for - I already expected rescue to be slower. Would love any suggestions to point me in a direction where I can get solid benchmarks. 

This branch:
```
Calculating -------------------------------------
with should_validate? method
                        46.000  i/100ms
-------------------------------------------------
with should_validate? method
                        479.288  (± 3.1%) i/s -      2.438k
```
Master with a rescue for ActiveRecord::RecordInvalid
```
Calculating -------------------------------------
with rescue ActiveRecord::RecordInvalid
                        43.000  i/100ms
-------------------------------------------------
with rescue ActiveRecord::RecordInvalid
                        435.397  (± 4.8%) i/s -      2.193k
```

eileencodes commented:
Ugg I do have vacation brain - forgot I had the `#save!` on the test scripts :grin:

Here are the benchmarks master vs my branch. There is a little bit of a slow down.

```
Calculating -------------------------------------
master
                        46.000  i/100ms
-------------------------------------------------
master
                        466.578  (± 5.1%) i/s -      2.346k
```

```
Calculating -------------------------------------
with should_validate?
                        44.000  i/100ms
-------------------------------------------------
with should_validate?
                        448.450  (± 5.6%) i/s -      2.244k
```

bgott commented:
Yo @tenderlove I ran the benchmarks against master and this branch. The standard deviation shows that there isn't much difference between the two branches. 

```
Calculating -------------------------------------
master
                        39.000  i/100ms
-------------------------------------------------
master
                        407.243  (± 4.4%) i/s -      2.067k


Calculating -------------------------------------
with should_validate?
                        40.000  i/100ms
-------------------------------------------------
with should_validate?
                        415.600  (± 3.4%) i/s -      2.080k
```

sgrif commented:
Thanks for running the benchmarks, guys. I just want to confirm, since nobody posted the script they used, this was benchmarking modifying existing records, not creating new ones, right?

eileencodes commented:
@sgrif I ran both and the results for the difference were the same

sgrif commented:
:+1:

On Sun, Feb 1, 2015, 7:24 PM Eileen M. Uchitelle <notifications@github.com>
wrote:

> @sgrif <https://github.com/sgrif> I ran both and the results for the
> difference were the same
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/18612#issuecomment-72398750>.
>

pacoguzman commented on activerecord/test/cases/validations/uniqueness_validation_test.rb:
```diff
@@ -386,4 +386,21 @@ def test_validate_uniqueness_on_empty_relation
+    topic.save!
+
+    assert_equal topic.replies.size, 1
+    assert reply.valid?
```
I have one question about this. Does this mean the reply will be valid from now on? I mean I thought calling valid? must execute the validations again

hundredwatt replied:
@pacoguzman It thought this was odd too. If you look at the new `should_validate?` method:
```
   def should_validate?(record) # :nodoc:
      !record.persisted? || record.changed? || record.marked_for_destruction?
    end
```
When `assert reply.valid?`, the record is [x] persisted, [x] un-changed, [x] not marked for destruction, so the uniqueness validation isn't performed. Before this PR was merged, the validation would have been run, but now it isn't.

hundredwatt replied:
As an example, if you change a different attribute, say `title`, then the validations will be run:

```
  def test_does_not_validate_uniqueness_of_if_parent_record_is_validate_false
    ...
    assert_equal topic.replies.size, 1
    assert reply.valid?
    reply.title = "Change An Attribute"
    assert reply.valid? #=> Fails!
    ...
  end
```

pacoguzman replied:
Understood thanks for the response. It's a big change for me. Then some
validations will run and others don't when the record has no changes.
Anyway I always add a condition on the uniqueness validations to avoid
extra queries
On Feb 7, 2015 5:59 PM, "Jason Nochlin" <notifications@github.com> wrote:

> In activerecord/test/cases/validations/uniqueness_validation_test.rb
> <https://github.com/rails/rails/pull/18612#discussion_r24290679>:
>
> > @@ -386,4 +386,21 @@ def test_validate_uniqueness_on_empty_relation
> >      topic = TopicWithUniqEvent.new
> >      assert topic.valid?
> >    end
> > +
> > +  def test_does_not_validate_uniqueness_of_if_parent_record_is_validate_false
> > +    Reply.validates_uniqueness_of(:content)
> > +
> > +    Reply.create!(content: "Topic Title")
> > +
> > +    reply = Reply.new(content: "Topic Title")
> > +    reply.save!(validate: false)
> > +    assert reply.persisted?
> > +
> > +    topic = Topic.new(reply_ids: [reply.id])
> > +    topic.save!
> > +
> > +    assert_equal topic.replies.size, 1
> > +    assert reply.valid?
>
> As an example, if you change a different attribute, say title, then the
> validations will be run:
>
>   def test_does_not_validate_uniqueness_of_if_parent_record_is_validate_false
>     ...
>     assert_equal topic.replies.size, 1
>     assert reply.valid?
>     reply.title = "Change An Attribute"
>     assert reply.valid? #=> Fails!
>     ...
>   end
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/18612/files#r24290679>.
>
